src/main/java/org/testng/JarFileUtils.java

```diff
@@ -52,23 +59,41 @@ private boolean testngXmlExistsInJar(File jarFile, List<String> classes) throws
         boolean foundTestngXml = false;
         try (JarFile jf = new JarFile(jarFile)) {
             Enumeration<JarEntry> entries = jf.entries();
+            File file = java.nio.file.Files.createTempDirectory("testngXmlPathInJar-").toFile();
+            String suitePath = null;
             while (entries.hasMoreElements()) {
                 JarEntry je = entries.nextElement();
-                if (matchesXmlPathInJar(je)) {
-                    Collection<XmlSuite> parsedSuites = Parser.parse(jf.getInputStream(je), processor);
-                    for (XmlSuite suite : parsedSuites) {
-                        // If test names were specified, only run these test names
-                        XmlSuite suiteToAdd = suite;
-                        if (testNames != null) {
-                            suiteToAdd = XmlSuiteUtils.cloneIfContainsTestsWithNamesMatchingAny(suite, testNames);
-                        }
-                        suites.add(suiteToAdd);
+                String jeName = je.getName();
+                if (Parser.canParse(jeName.toLowerCase())){ 
+                    InputStream  inputStream = jf.getInputStream(je);
+                    File copyFile = new File(file, jeName);
+                    Files.copyFile(inputStream, copyFile);
```

Comment:
You can use the jdk7 api for the copy: https://docs.oracle.com/javase/7/docs/api/java/nio/file/Files.html#copy(java.io.InputStream,%20java.nio.file.Path,%20java.nio.file.CopyOption...)

Reply:
Thanks for reminder, I tried this before, but this api doesn't support when parent folder is not existed.

Reply:
What about:
```java
Path rootDir = Files.createTempDirectory("testngXmlPathInJar-");
Path je = rootDir.resolve(jeName);
Files.copyFile(inputStream, je);
```


What about a `TestNameMatcher` with the suite as constructor param and a method: `XmlSuite

Reply:
It doesn't work.  je is only a path, if it doen't exist, you also have to create the folder. 
Not that neat as `org.testng.reporters.Files.copyFile` IMHO